pkg/storage/read_as_of_iterator.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package storage
+
+import "github.com/cockroachdb/cockroach/pkg/util/hlc"
+
+// ReadAsOfIterator wraps an iterator that implements the SimpleMVCCIterator
+// such that all MVCC keys returned are not delete tombstones and will have
+// timestamps less than the endtime timestamp, unless that key is invalid.
+type ReadAsOfIterator struct {
+	iter SimpleMVCCIterator
+
+	// endtime is the latest timestamp an MVCC key will have that the
+	// ReadAsOfIterator will return
+	endTime hlc.Timestamp
+
+	// valid is the latest boolean value returned by Valid()
+	valid bool
```

Comment:
I don't see why we need this field. All callers call `Valid()`, which sets this field as a side-effect, and then they read this field. Why don't the callers just read the return value of `Valid()` directly instead?

Reply:
Done.